library/ssl_msg.c

```diff
@@ -3106,32 +3106,127 @@ static size_t ssl_get_reassembly_buffer_size( size_t msg_len,
+        if( !( rret == 0 || rret == MBEDTLS_ERR_READER_DATA_LEFT ) )
+        {
+            ret = rret;
+        }
+    }
+
+    return( ret );
+}
+
+static int ssl_defrag_accumulate( mbedtls_ssl_context const *ssl )
+{
+    int ret;
+    mbedtls_ssl_hs_defrag *defrag = &ssl->handshake->defrag;
+
+    ret = mbedtls_reader_feed( defrag->reader, ssl->in_msg, ssl->in_msglen );
+
+    MBEDTLS_SSL_DEBUG_RET( 1, ( "mbedtls_reader_feed" ), ret );
+
+    if( ret == MBEDTLS_ERR_READER_NEED_MORE )
+    {
+        /* This error code indicates that the reader is still
+         * unable to satisfy the last `mbedtls_reader_get` request.
+         *
+         * For example, this will happen if the handshake message
+         * has been split into several fragments, and the last
+         * fragment is not yet available.
+         *
+         * In this situation, we need to continue to read additional
+         * TLS records and feed their contents to the reader.
+         *
+         * We accomplish this by re-routing the error code into
+         * `MBEDTLS_ERR_SSL_CONTINUE_PROCESSING` and looping.
+         */
+        MBEDTLS_SSL_DEBUG_MSG( 2, ( "Reader needs more data" ) );
+
+        ret = MBEDTLS_ERR_SSL_CONTINUE_PROCESSING;
+    }
+
+    return( ret );
+}
+#endif /* MBEDTLS_SSL_HS_DEFRAG_CLT */
+
 static uint32_t ssl_get_hs_total_len( mbedtls_ssl_context const *ssl )
 {
-    return( ( ssl->in_msg[1] << 16 ) |
-            ( ssl->in_msg[2] << 8  ) |
-              ssl->in_msg[3] );
+    unsigned char *hs_msg = mbedtls_ssl_hs_msg_ptr( ssl );
+    return( ( hs_msg[1] << 16 ) |
+            ( hs_msg[2] << 8  ) |
+              hs_msg[3] );
 }
 
 int mbedtls_ssl_prepare_handshake_record( mbedtls_ssl_context *ssl )
 {
+    int ret;
+    const unsigned char *hs_msg;
+
+#if defined(MBEDTLS_SSL_HS_DEFRAG_CLT)
+    if( mbedtls_ssl_hs_defrag_enabled( ssl ) )
+    {
+        if( ( ret = ssl_defrag_try_read( ssl, mbedtls_ssl_hs_hdr_len( ssl ), 0 ) ) != 0 )
+            return( ret );
+    }
+    else
+#endif /* MBEDTLS_SSL_HS_DEFRAG_CLT */
     if( ssl->in_msglen < mbedtls_ssl_hs_hdr_len( ssl ) )
     {
         MBEDTLS_SSL_DEBUG_MSG( 1, ( "handshake message too short: %d",
                             ssl->in_msglen ) );
         return( MBEDTLS_ERR_SSL_INVALID_RECORD );
     }
 
+    hs_msg = mbedtls_ssl_hs_msg_ptr( ssl );
     ssl->in_hslen = mbedtls_ssl_hs_hdr_len( ssl ) + ssl_get_hs_total_len( ssl );
 
     MBEDTLS_SSL_DEBUG_MSG( 3, ( "handshake message: msglen ="
                         " %d, type = %d, hslen = %d",
-                        ssl->in_msglen, ssl->in_msg[0], ssl->in_hslen ) );
+                        ssl->in_msglen, hs_msg[0], ssl->in_hslen ) );
 
 #if defined(MBEDTLS_SSL_PROTO_DTLS)
     if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
     {
-        int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+        ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
         unsigned int recv_msg_seq = ( ssl->in_msg[4] << 8 ) | ssl->in_msg[5];
```

Comment:
Should those uses of `ssl->in_msg` be replaced by `hs_msg`, too?

Reply:
If I am not mistaken, the sequence number is part of the header of the DTLS handshake message.  If this is indeed the case, then these uses should be replaced by `hs_hdr` vs `hs_msg`. I will take an extra look.